### PR TITLE
Make optional.Option[T] type serializable

### PR DESCRIPTION
**Pull request by 6543**
make the generic `Option` type de-/serializable for json and yaml

### CODE REVIEW

File: modules/optional/serialization.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		return err
+	}
+	if v == nil {
+		return nil
+	}
+	var someValue T
+	if err := json.Unmarshal(data, &someValue); err != nil {
+		return err
+	}
+	*o = Some[T](someValue)
```
Comment (KN4CK3R) [inline modules/optional/serialization.go]:
```suggestion
	var v *T
	if err := json.Unmarshal(data, &v); err != nil {
		return err
	}
	*o = FromPtr(v)
```

File: modules/optional/serialization.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		return err
+	}
+	if v == nil {
+		return nil
+	}
+	var someValue T
+	if err := value.Decode(&someValue); err != nil {
+		return err
+	}
+	*o = Some[T](someValue)
```
Comment (KN4CK3R) [inline modules/optional/serialization.go]:
```suggestion
	var v *T
	if err := value.Decode(&v); err != nil {
		return err
	}
	*o = FromPtr(v)
```

**Comment by lunny** [comment]
This PR should be merged first then #29329

**Comment by 6543** [comment]
They are related but independent :)

As the changed struct in mentioned pull are not used for de-/serializion

**Comment by 6543** [comment]
PS: I'm happy to see it merged once it has sufficient reviews ;)

File: modules/optional/serialization.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	if err := json.Unmarshal(data, &v); err != nil {
+		return err
+	}
+	*o = FromPtr(v)
+	return nil
+}
+
+func (o Option[T]) MarshalJSON() ([]byte, error) {
+	if !o.Has() {
+		return []byte("null"), nil
```
Comment (delvh) [inline modules/optional/serialization.go:23]: So `null` is set in JSON but not in YAML?

**Review by delvh — APPROVED** [review]

File: modules/optional/serialization.go (hunk 3 above)
Comment (6543) [reply modules/optional/serialization.go:23]: Yes but output is the same, see tests

**Comment by KN4CK3R** [comment]
When I added the option type I had a look at the chi binding which is needed for form binding but that looked more complicated. But we should add support there too before we start to use the option type in forms.

**Comment by 6543** [comment]
@KN4CK3R yes but that's outside of this pull scope i would say ?

**Comment by KN4CK3R** [comment]
yes, just want to share my thoughts.

File: modules/optional/serialization.go (hunk 3 above)
Comment (6543) [reply modules/optional/serialization.go:23]: json serializer expect to get the string returned to print as serialized value.
yaml serializer expect to get a type returned (mostly some basic type) to finish serialization

File: modules/optional/serialization_test.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+			obj:  new(testSerializationStruct),
+			want: `{"normal_string":"","normal_bool":false,"optional_two_bool":null,"optional_twostring":null}`,
+		},
+		{
+			name: "some",
+			obj: &testSerializationStruct{
+				NormalString: "a string",
+				NormalBool:   true,
+				OptBool:      optional.Some(false),
+				OptString:    optional.Some(""),
```
Comment (KN4CK3R) [inline modules/optional/serialization_test.go:43]:
```suggestion
				OptString:    optional.Some(""),
				OptTwoBool:   optional.None[bool](),
				OptTwoString: optional.None[string](),
```
Should we add the default values here too? I spend more time than needed to understand why `"optional_two_bool":null,"optional_twostring":null` results in `false` and `""`.

Comment (6543) [reply modules/optional/serialization_test.go:43]: @KN4CK3R that's intentional as i have set `omitempty` tag for Opt* but NOT for OptTwo*

Comment (6543) [reply modules/optional/serialization_test.go:43]: > why "optional_two_bool":null,"optional_twostring":null results in false and "".

oh right that one it just **does not** you might got iritated by the "Normal" types that are in there to cross check

Comment (KN4CK3R) [reply modules/optional/serialization_test.go:43]: yes, just a visual problem not a code problem.

**Comment by KN4CK3R** [comment]
> PS: `optional.None[bool]()` should be equivalent to not set them, else it's a bug in the optional.Option type itself

It is identical.

**Review by KN4CK3R — APPROVED** [review]

**Pull request merged**